### DISCUSSION TITLE
How to export all bookmarks/outlines and import them

### DISCUSSION THREAD

**Discussion by dl0120** [py-pdf/pypdf#1625] [Q&A] [answered]
The problem: I have a PDF file with nested bookmarks. When I process it, I lose the bookmarks. I want to save the nested bookmark information before processing, and then put in the same bookmarks after processing.

What I've done: I looked at the documentation, and source code. I started playing around with it, and realized that PdfReader's outline property gives me a nested list of dictionary for outline item and list for nesting of outline items. I tried using PdfWriter's add_outline_item_dict and couple other methods. Unfortunately, I get an error - it seems like I can can only give it one outline item at a time. So I was thinking maybe I need to write my own function. But before doing that I wanted to browse the code a bit to see if I can avoid reinventing the wheel. I downloaded github desktop etc. etc. and now I am here.

Thank you!

**Comment by pubpub-zz** [comment, marked as answer]
Normally if you are using use `append()` with `import_outline=True` you should  keep all the outlines that applies to the imported pages. have you tried it?

**Reply by dl0120** [reply]
Thank you for the reply! Right now it's not the merging that is losing the bookmarks - it's the processing with other programs that are not pypdf (work related) :(

**Reply by MartinThoma** [reply]
If it's not pypdf causing the issue, what do you expect from us?

**Reply by dl0120** [reply]
Hi, thank you for the reply. I was wondering if someone could let me know pypdf had a built-in way to import/export bookmarks so that I could avoid reinventing the wheel if I didn't have to. I wasn't implying that pypdf was defective or causing any issues.

**Reply by MartinThoma** [reply]
Is https://stackoverflow.com/a/63519098/562769 what you're looking for?

However, that flattens the bookmarks. But it might make is easier for you to handle.

If you don't want it flattened, you have to write custom code around reader.outlines.

**Comment by pubpub-zz** [comment]
`reader.outline` will return with list of dict/list. list will correspond to lower level. 
ex to see them:
```python
def p(a, pref=""):
    if isinstance(a, list):
        for aa in a:
            p(aa, pref + " ")
    else:
        print(pref + a["/Title"])


p(r.outline)
```

The dict will provide you with all the required fields. Also `.indirect_reference` will provide you with the original outline object
you should be able to recreate them with the `.add_outline_item_dict()`

**Comment by Firestar-Reimu** [comment]
can pypdf export the destination point (**position on target page**) of bookmarks?

for example, the `get_toc` of pymupdf return to: **position on target page** ([Point](https://pymupdf.readthedocs.io/en/latest/point.html#point)). Point=0 is top of the page

https://pymupdf.readthedocs.io/en/latest/document.html#Document.get_toc

**Reply by stefan6419846** [reply]
I recommend you to have a look at the PDF 2.0 specification (ISO 32000-2:2020): The relevant definitions are in section 12.3.3 which in table 151 details the entries of each outline item. Its `/Dest` entries provides information on the targets. Non-named links might carry additional positioning commands as detailed in table 149 of the PDF 2.0 specification (ISO 32000-2:2020). In general, section 12.3.2 and its references to other sections are relevant for how targets are defined. https://github.com/py-pdf/pypdf/discussions/2911 might help as well for mapping page numbers - while being about extracting internal links, the basic structure is similar.
